gsi_openssh/source/fips_mode_replacement.h

```diff
@@ -16,10 +16,9 @@
 
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
 /*
- * OpenSSL version 3.0 and up no longer has FIPS_mode().
- * Making a replacement function is not feasible since FIPS would need to be
- * initialized differently in any case.
- * See https://www.openssl.org/docs/manmaster/man7/fips_module.html for details
+ * OpenSSL versions 3.0 and up no longer have FIPS_mode(). To support both
+ * OpenSSL 3.x and older versions for other OSes, we use the replacement
+ * function as shipped by Fedora/RHEL/CentOS in their OpenSSL 3.x packages.
  */
-# define FIPS_mode() 0
-#endif
\ No newline at end of file
+# define FIPS_mode() EVP_default_properties_is_fips_enabled(NULL)
```

Comment:
I think we should protect this with an `#ifndef` **_inside_** the `#if OPENSSL_VERSION_NUMBER >= 0x30000000L`, something like
```
#ifndef FIPS_mode
# define FIPS_mode()   EVP_default_properties_is_fips_enabled(NULL)
#endif
```
in case we get it also from somewhere else.

Reply:
But do we want that? Without the `#ifndef` we know what is defined during compilation in all files that use `FIPS_mode()`. When we depend on some possible other definition we can't know that.

Reply:
But if it's already defined (for example by Fedora or so), it will break compilation. You mean you want to force breaking compilation such that it can manually be removed? Even in that case you still want to protect against breaking when the header is included twice.
If you like you could always add a `#warning` about it already being defined?

Reply:
I see, what about:
```
#ifdef FIPS_mode
#undef FIPS_mode
#endif
# define FIPS_mode()   EVP_default_properties_is_fips_enabled(NULL)
```
...then - man that was hard to write down thanks to the autocomplete of the editor here on GitHub always replacing things after hitting return.

Reply:
that would work and is in principle fine with me, but why do we need to use our own definition? After all, it's only as a workaround for when someone else hasn't provided it yet?
In any case, it's probably best to put in the header the typical "headerfile protection", so start the header file with
```
#ifndef FIPS_MODE_REPLACEMENT_H
#define FIPS_MODE_REPLACEMENT_H
```
and end with
```
#endif /* FIPS_MODE_REPLACEMENT_H */
```
such that at least the `#undef` and `#define` of `FIPS_mode()` will be done only once even when the header is included multiple times.

Reply:
> that would work and is in principle fine with me, but why do we need to use our own definition? After all, it's only as a workaround for when someone else hasn't provided it yet?

Yeah, I think you're right. My thinking was that we would then know what the user would have included, but we can't be sure anyhow, so what.

Hence I'll just add the "headerfile protection" and `#ifndef FIPS_mode[...]#endif` around the definition of `FIPS_mode()`.